File: src/main/scala/resonantinduction/core/prefab/imprint/TileFilterable.java
package resonantinduction.core.prefab.imprint;

import java.util.Set;

import net.minecraft.block.material.Material;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.common.ForgeDirection;
import resonant.api.IFilterable;
import resonant.api.IRotatable;
import resonant.lib.content.module.prefab.TileInventory;
import universalelectricity.api.vector.Vector3;

public abstract class TileFilterable extends TileInventory implements IRotatable, IFilterable
{
	private ItemStack filterItem;
	private boolean inverted;
	public static final int FILTER_SLOT = 0;
	public static final int BATERY_DRAIN_SLOT = 1;

	public TileFilterable()
	{
		super(null);
		this.maxSlots = 2;
	}

	public TileFilterable(Material material)
	{
		super(material);
		this.maxSlots = 2;
	}

	protected boolean isFunctioning()
	{
		return true;
	}

	/** Allows filters to be placed inside of this block. */
	@Override
	public boolean use(EntityPlayer player, int side, Vector3 hit)
	{
		ItemStack containingStack = getFilter();

		if (containingStack != null)
		{
			if (!world().isRemote)
			{
				EntityItem dropStack = new EntityItem(world(), player.posX, player.posY, player.posZ, containingStack);
				dropStack.delayBeforeCanPickup = 0;
				world().spawnEntityInWorld(dropStack);
			}

			setFilter(null);
			return true;
		}
		else
		{
			if (player.getCurrentEquippedItem() != null)
			{
				if (player.getCurrentEquippedItem().getItem() instanceof ItemImprint)
				{
					setFilter(player.getCurrentEquippedItem());
					player.inventory.setInventorySlotContents(player.inventory.currentItem, null);
					return true;
				}
			}
		}

		return false;
	}

	@Override
	public boolean configure(EntityPlayer player, int side, Vector3 hit)
	{
		toggleInversion();
		markUpdate();
		markRender();
		return true;
	}

	/**
	 * Looks through the things in the filter and finds out which item is being filtered.
	 * 
	 * @return Is this filterable block filtering this specific ItemStack?
	 */
	public boolean isFiltering(ItemStack itemStack)
	{
		if (this.getFilter() != null && itemStack != null)
		{
			Set<ItemStack> checkStacks = ItemImprint.getFilters(getFilter());

			if (checkStacks != null)
			{
				for (ItemStack stack : checkStacks)
				{
					if (stack.isItemEqual(itemStack))
					{
						return !inverted;
					}
				}
			}
		}

		return inverted;
	}

	@Override
	public void setFilter(ItemStack filter)
	{
		this.filterItem = filter;
		this.worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
	}

	@Override
	public ItemStack getFilter()
	{
		return this.filterItem;
	}

	public void setInverted(boolean inverted)
	{
		this.inverted = inverted;
		this.worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
	}

	public boolean isInverted()
	{
		return this.inverted;
	}

	public void toggleInversion()
	{
		setInverted(!isInverted());
	}

	@Override
	public ForgeDirection getDirection()
	{
		return ForgeDirection.getOrientation(getBlockType() != null ? getBlockMetadata() : 0);
	}

	@Override
	public void setDirection(ForgeDirection facingDirection)
	{
		this.worldObj.setBlockMetadataWithNotify(this.xCoord, this.yCoord, this.zCoord, facingDirection.ordinal(), 3);
	}

	@Override
	public void writeToNBT(NBTTagCompound nbt)
	{
		super.writeToNBT(nbt);
		nbt.setBoolean("inverted", inverted);
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		if (nbt.hasKey("filter"))
		{
			this.getInventory().setInventorySlotContents(0, ItemStack.loadItemStackFromNBT(nbt.getCompoundTag("filter")));
		}
		inverted = nbt.getBoolean("inverted");
	}

}

File: src/main/scala/edx/basic/blocks/TileFilterable.scala
package edx.basic.blocks

import net.minecraft.block.material.Material
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.util.ForgeDirection
import resonant.api.tile.{IFilterable, IRotatable}
import resonant.lib.prefab.tile.mixed.TileInventory
import resonant.lib.transform.vector.Vector3

object TileFilterable
{
  final val FILTER_SLOT: Int = 0
  final val BATERY_DRAIN_SLOT: Int = 1
}

abstract class TileFilterable extends TileInventory(Material.wood) with IRotatable with IFilterable
{
  private var filterItem: ItemStack = null
  private var inverted: Boolean = false

  override def getSizeInventory: Int = 2

  /**
   * Allows filters to be placed inside of this block.
   */
  override def use(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    val containingStack: ItemStack = getFilter
    if (containingStack != null)
    {
      if (!world.isRemote)
      {
        val dropStack: EntityItem = new EntityItem(world, player.posX, player.posY, player.posZ, containingStack)
        dropStack.delayBeforeCanPickup = 0
        world.spawnEntityInWorld(dropStack)
      }
      setFilter(null)
      return true
    }
    else
    {
      if (player.getCurrentEquippedItem != null)
      {
        if (player.getCurrentEquippedItem.getItem.isInstanceOf[ItemImprint])
        {
          setFilter(player.getCurrentEquippedItem)
          player.inventory.setInventorySlotContents(player.inventory.currentItem, null)
          return true
        }
      }
    }
    return false
  }

  def setFilter(filter: ItemStack)
  {
    this.filterItem = filter
    this.worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
  }

  def getFilter: ItemStack =
  {
    return this.filterItem
  }

  override def configure(player: EntityPlayer, side: Int, hit: Vector3): Boolean =
  {
    toggleInversion
    markUpdate
    markRender
    return true
  }

  def toggleInversion
  {
    setInverted(!isInverted)
  }

  def isInverted: Boolean =
  {
    return this.inverted
  }

  def setInverted(inverted: Boolean)
  {
    this.inverted = inverted
    this.worldObj.markBlockForUpdate(xCoord, yCoord, zCoord)
  }

  /**
   * Looks through the things in the filter and finds out which item is being filtered.
   *
   * @return Is this filterable block filtering this specific ItemStack?
   */
  def isFiltering(itemStack: ItemStack): Boolean =
  {
    if (this.getFilter != null && itemStack != null)
    {
      val checkStacks: java.util.List[ItemStack] = ItemImprint.getFilters(getFilter)
      if (checkStacks != null)
      {
        import scala.collection.JavaConversions._
        for (stack <- checkStacks)
        {
          if (stack.isItemEqual(itemStack))
          {
            return !inverted
          }
        }
      }
    }
    return inverted
  }

  override def getDirection: ForgeDirection =
  {
    return ForgeDirection.getOrientation(if (getBlockType != null) getBlockMetadata else 0)
  }

  override def setDirection(facingDirection: ForgeDirection)
  {
    this.worldObj.setBlockMetadataWithNotify(this.xCoord, this.yCoord, this.zCoord, facingDirection.ordinal, 3)
  }

  override def writeToNBT(nbt: NBTTagCompound)
  {
    super.writeToNBT(nbt)
    nbt.setBoolean("inverted", inverted)
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)
    if (nbt.hasKey("filter"))
    {
      this.getInventory.setInventorySlotContents(0, ItemStack.loadItemStackFromNBT(nbt.getCompoundTag("filter")))
    }
    inverted = nbt.getBoolean("inverted")
  }

  protected def isFunctioning: Boolean =
  {
    return true
  }

}